java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -172,35 +172,64 @@ RelationalSymbolicValue inverse() {
     return results;
   }
 
+  private boolean hasBooleanSymbolicValueAsOperand() {
+    return (leftOp instanceof BooleanSymbolicValue) || (rightOp instanceof BooleanSymbolicValue);
+  }
+
+  private boolean isEqualityCheck() {
+    return Kind.METHOD_EQUALS == kind || Kind.NOT_METHOD_EQUALS == kind || Kind.EQUAL == kind || Kind.NOT_EQUAL == kind;
+  }
+
   private List<ProgramState> addNullConstraintsForBooleanWrapper(ProgramState initialProgramState, List<ProgramState> copiedConstraints) {
     BooleanConstraint leftConstraint = initialProgramState.getConstraint(leftOp, BooleanConstraint.class);
     BooleanConstraint rightConstraint = initialProgramState.getConstraint(rightOp, BooleanConstraint.class);
-    if (leftConstraint != null && rightConstraint == null && !isEquality()) {
-      List<ProgramState> nullConstraints = copiedConstraints.stream()
-        .flatMap(ps -> rightOp.setConstraint(ps, ObjectConstraint.NULL).stream())
-        .map(ps -> ps.removeConstraintsOnDomain(rightOp, BooleanConstraint.class)
-      ).collect(Collectors.toList());
-      return ImmutableList.<ProgramState>builder().addAll(copiedConstraints).addAll(nullConstraints).build();
+    
+    if (shouldAddNullConstraint(leftConstraint, rightConstraint)) {
+      return getProgramStatesWithNullConstraints(copiedConstraints, leftOp);
+    }
+    
+    if (shouldAddNullConstraint(rightConstraint, leftConstraint)) {
+      return getProgramStatesWithNullConstraints(copiedConstraints, rightOp);
     }
     return copiedConstraints;
   }
 
+  private boolean shouldAddNullConstraint(@Nullable BooleanConstraint constraint1, @Nullable BooleanConstraint constraint2) {
+    return !isEquality() && constraint1 == null && constraint2 != null;
+  }
+
+  private static List<ProgramState> getProgramStatesWithNullConstraints(List<ProgramState> copiedConstraints, SymbolicValue operand) {
+    List<ProgramState> nullConstraints = copiedConstraints.stream()
+      .flatMap(ps -> operand.setConstraint(ps, ObjectConstraint.NULL).stream())
+      .map(ps -> ps.removeConstraintsOnDomain(operand, BooleanConstraint.class)
+      ).collect(Collectors.toList());
```

Comment:
Question of style, but I would really prefer to have the parenthesis on the previous line and this one starting with `.collect(` to align dots